File: deck.py
import random
from card import Card

class Deck:
    def __init__(self):
        self._cards = []
        self._dealt = []
        self._discarded = []
        self.populate()
    
    def populate(self):
        self._cards = [Card(n, s) for s in Card.suits for n in Card.ranks]
    
    def shuffle(self):
        self._cards.extend(self._discarded)
        self._cards.extend(self._dealt)
        self._discarded = []
        self._dealt = []
        random.shuffle(self._cards)
    
    def deal(self):
        new_card = self._cards.pop()
        self._dealt.append(new_card)
        return new_card
    
    def discard(self):
        self._discarded.extend(self._dealt)
        self._dealt = []

File: src/js/deck.js
import Card, {RANKS, SUITS} from './card.js';

/**
 * Represents a deck of playing cards.
 */
class Deck {
    /**
     * Represents a deck of cards.
     * @constructor
     */
    constructor() {
        this.cards = [];
        this.dealt = [];
        this.discarded = [];
        this.populate();
    }

    /**
     * Populates the deck with a standard set of playing cards.
     */
    populate() {
        this.cards = [];
        this.dealt = [];
        this.discarded = [];
        for (let suit of SUITS) {
            for (let rank of RANKS) {
                this.cards.push(new Card(rank, suit));
            }
        }
    }

    /**
     * Shuffles the deck by randomly reordering the cards.
     */
    shuffle() {
        this.cards = this.cards.concat(this.discarded);
        this.discarded = [];
        for (let i = this.cards.length - 1; i > 0; i--) {
            let j = Math.floor(Math.random() * (i + 1));
            let temp = this.cards[i];
            this.cards[i] = this.cards[j];
            this.cards[j] = temp;
        }
    }

    /**
     * Draws a card from the deck.
     * 
     * @param {boolean} [faceDown=false] - Whether the drawn card should be face down.
     * @returns {object} - The drawn card.
     */
    draw(faceDown = false) {
        if (this.cards.length === 0) {
            this.shuffle();
        }
        let card = this.cards.pop();
        card.faceUp = !faceDown;
        this.dealt.push(card);
        console.log('draw card', card);
        return card;
    }

    /**
     * Discards the dealt cards and adds them to the discarded pile.
     */
    discard() {
        this.discarded = this.discarded.concat(this.dealt);
        this.dealt = [];
    }
}

export default Deck;